### DISCUSSION TITLE
Welcome to mh-z19 Discussions!

### DISCUSSION THREAD

**Discussion by UedaTakeyuki** [UedaTakeyuki/mh-z19#31] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ehyangg** [comment]
![cxzczx](https://user-images.githubusercontent.com/93096579/138609211-1c63fa00-a322-4839-b404-b437a319228a.PNG)

1. When I write a script like a picture, the terminal only blinks.
I am using raspi4 and python3.7.
I don't know much about Python and Raspberry pi, but can you tell me kindly?

**Reply by ehyangg** [reply]
i solved this problem.

**Comment by ehyangg** [comment]
what is "TT, SS, UhUl" ?? what is meaning??

**Reply by UedaTakeyuki** [reply]
refer [6.10. Command 0x86 (read concentration)](https://revspace.nl/MHZ19#Setting_the_measurement_range)

**Comment by false-git** [comment]
すいません、日本語で書いても大丈夫でしょうか？駄目ならば書き直します。

今まで、Raspberry pi 3B でうまく動いていたのですが、β版の64bit OS を入れたらたまに応答電文の受信に失敗するようになりました。
通常時は、以下のような電文が帰ってきます。
`ff 86 03 cb 43 00 29 00 40`
しかし、たまにffと86の間に00が挟まった以下のような電文が帰ってきます。
`ff 00 86 03 cb 43 00 29 00`
このとき、read_all() 関数が dict ではなく 9 を返します。(ser.writeの結果を返すため)

この問題に対応するために、チェックサムのチェックをして、あっていなかったら3回リトライするように修正したコードを
https://github.com/false-git/mh-z19/commit/db177d484da448b8874a452129e8d0246ebbe43f
に置きました。

ここから、Pull Request を出しても良いものでしょうか？

**Reply by UedaTakeyuki** [reply]
No problem, I don' mind Japanese comments.
All pull requests are really welcome!

But wait a minute.
I've watched your code, I **don't really see** the necessity of the function ``convert_telegram``.  The words ``telegram`` also sounds weird.

I.M.H.O, How about moving checksum comparison into the current condition just looking  **length and firstword** of the response.

```if len(s) >= 9 and s[0] == 0xff and s[1] == 0x86:```

**Reply by false-git** [reply]
> I don't really see the necessity of the function `convert_telegram`.

この関数で、python2 と python 3 の差を吸収しているつもりです。python2 の str を int のリストに変換します。

> The words telegram also sounds weird.

`convert_response` なら良いでしょうか？

> How about moving checksum comparison into the current condition just looking length and firstword of the response.

すいません、意味が取り切れていないのですが、lengthと 0xff だけチェックして、0x86 のチェックはチェックサムのみで良いのではないかと言う意味ですか？

**Reply by UedaTakeyuki** [reply]
I think the lack of checksum comparison of current logic is caused by my skimpy coding :-), so both p2 and p3 need to add checksum comparison. No need to hesitate to affect both.

> すいません、意味が取り切れていないのですが、lengthと 0xff だけチェックして、0x86 のチェックはチェックサムのみで良いのではないかと言う意味ですか？

Exactly!

**Comment by ehyangg** [comment]
I'm coding mh-z19, and I'm going to measure the carbon dioxide concentration in real time and code it so that it can make a sound through buzzer when it exceeds the reference value. Is there a way to do it in real time? I think you know.

**Reply by UedaTakeyuki** [reply]
Currently, the response time improved version (about 300 ms) is developing as a product.

**Comment by cbzihou** [comment]
Hello, I am learning about your work and would like to ask:
I am not connected to the TEMPERATURE and humidity sensor. Why does the temperature and humidity data appear and change?

1. Your instructions 
![0](https://user-images.githubusercontent.com/94046319/144021533-dab669c6-c3c8-4b10-82c7-57d13a8d7ae9.png)

2. my
![1](https://user-images.githubusercontent.com/94046319/144021572-72998c4c-d667-4d9c-83a0-459cd74fcebd.png)

**Reply by UedaTakeyuki** [reply]
Hi cbzihou, thank you for your using!

The 'temperature' value you mentioned is thought as related to "internal" ambient temperature which the sensor uses for calculating the CO2 value.

For more detail, you can refer https://github.com/UedaTakeyuki/mh-z19#undocumented-response-values-of-0x86-command.

By the way, I'm afraid that there is no humidity in the return values.

**Comment by jonesthefox** [comment]
Hello doctor Ueda

I have a question about the mh-z19c. It features an red LED inside the sensor, that (in my case) keeps blinking. 
Is that some kind of status LED or is it a byproduct of the measuring? As far as I understand, it uses infrared for the measurement, so i should not be able see it..

But when it's just for status, is it possible to disable it? :)

**Comment by UedaTakeyuki** [comment]
To say frankly that I don't know the detail, but I think it is the latter because there is no description in the [data sheet](https://www.winsen-sensor.com/d/files/MH-Z19B.pdf).
I also welcome more detailed information about it, thank you.

**Reply by jonesthefox** [reply]
I took the chance and wrote a mail to winsen :)

Im really not 100% sure, and unfortunately can't recall the source, but i believe, i was reading something somewhere about a status led when i searched the internet as i was evaluating the sensor for my use.
But maybe i'm just crossreferencing one of the 1000 tutorials for air quality testers :)

Anyway, i'll let you know what they answer!

And thank you for your good work on the mh-z19 module, it's really useful.

**Reply by UedaTakeyuki** [reply]
I'm looking forward to your answer! Thank you jonesthefox!

**Reply by jonesthefox** [reply]
Unfortunately it's not possible to turn the led off.

screenshot:
![E5455E26-47ED-4069-B8E8-43185215F0D6](https://user-images.githubusercontent.com/10356409/150776731-083817eb-0a04-45a0-868f-2af109f3dd06.jpeg)

Now we know :-)

**Reply by UedaTakeyuki** [reply]
Thank you jonesthefox!

**Comment by ghost** [comment]
Dear Doctor Ueda,
I hope this is the right place to ask. If not, please forgive and kindly forward my issue to the correct forum.
I'm working with the new model mh-z19C and the raspberry pi. 
Unfortunately, I can't get the UART connection to work. I tried with 2 separate sensors and 2 raspberry pis (1x raspi 2b, 1x raspi4).
I successfully enabled UART (I see the serial devices using ls -la /dev and UART=1 in the config.txt). However, when running "python -m mh_z19", I only receive an empty json { }. I double and triple checked the wiring on both raspberries.
Then running the sensor using PWM, it is working fine, I receive a value, the three orange LEDs are blinking every half second or so.
I looked at the datasheet of the C version, the TX and RX PINs seem to be identical to the B version. 
Could you kindly confirm that your code is also working with the C version ? 
Many thanks for your ongoing support and kind regards from Austria,
Stefan

**Reply by jonesthefox** [reply]
I hope nobody minds if I answer in german :)

Hi Stefan

Ich habe auch den MH-Z19C in betrieb, via UART an einem Pi3. 

- Hast du einmal getestet ob es funktioniert, wenn du 
`python mh_z19.py --serial_console_untouched`
aufrufst? (Wenn du diesen Parameter weg lässt, fragt er dich nach deinem Passwort für sudo und versucht ein eventuell laufendes getty zu beenden.) 

- Gruppenzugehörigkeit deines Users geprüft, ob er das serial device nutzen darf? 

- Wenn du deine verkabelung doppelt und dreifach überprüft hast, frage ich nicht ob du vielleicht tx mit tx und rx mit rx verbunden hast... ;)

Gruss aus der Schweiz! 
^^

**Comment by UedaTakeyuki** [comment]
Hi Stefan,

I don't try it but I've heard some reports that the type C version of mh_z19 is working well with UART as other type B and A.

**Comment by ghost** [comment]
Good Morning to Japan and CH, many thanks for your replies... 
I did some more testing on the 2 sensors I bought using an Arduino. Result is the same - I don't get any values from the UART, only from the PWM, and they kind of deviate a lot between the 2 sensors (they are kind of similar only when using 2000 vs 5000 as ranges). 
I then compared in more detail pictures and fritzing diagrams from other users, and they don't match at all in terms of layout- no idea what I bought in China,.. but I'm pretty sure they are a fake.
Sorry for wasting your time... will run some additional tests using PWM and then probably end up buying a real one on Amazon... :-)
Kind regards,
Stefan

**Reply by jonesthefox** [reply]
When i was evaluating this sensor i stumbled over some sites with infos about the faked sensors:
[https://revspace.nl/MH-Z19B](https://revspace.nl/MH-Z19B) and [https://emariete.com/en/co2-mh-z19b-fake-sensors-2/](https://emariete.com/en/co2-mh-z19b-fake-sensors-2/)

seems the casings are different and the pcb is black..

**Comment by UedaTakeyuki** [comment]
Hi Stefan,

I've heard that some of these sensor modules have a low signal level. In case, [pull up](https://github.com/UedaTakeyuki/mh-z19#in-case-you-cant-get-the-value) may one of the possible way to fix.

**Comment by ghost** [comment]
Hello, thanks for these hints, I found them as well... (I read extensively all threads and comments to not come up with an issue already discussed)... I should have checked the pics and comments before buying these assumingly fakes on EBay (this would be the link btw: https://www.ebay.de/itm/393917199637, however mine look different on the back, with no correct printing of the 9 pins).
Many thanks for your support. I will experiment with the PWM signal, the reading seems roughly accurate, and for getting to my goal (integrating the sensor data into HomeAssistant via MQTT) this is more than enough. 
Kind regards once again,
Stefan

**Comment by tomoher** [comment]
Hi,
I can use the mh_z19.read_all() command to get data from the sensor in a python script - great!

I'm not sure which pondslider command to use to save the returned value into a txt file. Any guidance, please?

Thank you!
Tomi

**Reply by UedaTakeyuki** [reply]
Hi Tomi, 
Thank you for using this library.

To save value into text file, you can use ``handlers.value.saver.save2strage``.
At default, your ``config.toml`` file on the mh_z19 folder might be as follow:

```
[[sensors]]
  handler = "mh_z19"
  [[sensors.values]]
    name = "co2"
    handlers = [
      "send2monitor", # with send2monitor.ini on the current directory.
#      "handlers.value.sender.send2m2x", # with send2m2x.ini on the handler/value/sender/send2m2x
#      "handlers.value.saver.save2strage"
  ]
```

So, check out the # from the last element of the **handlers** array as follow.

```
[[sensors]]
  handler = "mh_z19"
  [[sensors.values]]
    name = "co2"
    handlers = [
      "send2monitor", # with send2monitor.ini on the current directory.
#      "handlers.value.sender.send2m2x", # with send2m2x.ini on the handler/value/sender/send2m2x
      "handlers.value.saver.save2strage"
  ]
```

Then, restart service as follow

```
autostart.sh --off
autostart.sh --on
```

Then, ``DATA/Value`` folder must be created on the /home/pi folder.
In this folder, the **co2.csv** must be created.

For the configuration of ``handlers.value.saver.save2strage``, please refer ``handlers/value/saver/save2strage/save2strage.ini``.

I should have written more detail document about it. Sorry for bothering you.

**Comment by GiChoc** [comment]
Hello, 

I hope you are well. 

I've got two main issues with making the mh_z19.py to work: 

1. Even if though I have _chmod 666 /dev/ttyS0_ and _chmod 666 /dev/serial0_ , when running  mh_z19.py, first time it works and then it gives me permission denied for /dev/serialS0. If I chmod 666 again, then it works. 
2. At mh_z19.py line 70, the condition _if len(s) >= 4 and s[0] == 0xff and s[1] == 0x86 and ord(checksum(s[1:-1])) == s[-1]:_ never becomes true, hence I never get a value. The only way to make this work is to remove the if in line 70. However, the values I get then are always around 2500 which seems too high. 

Thank you for your help in advance!

**Reply by UedaTakeyuki** [reply]
1. try ```read(True)``` instead of ```read()```
2. Might be a hardware problem of the sensor module. Refer https://github.com/UedaTakeyuki/mh-z19/issues/52.

**Reply by GiChoc** [reply]
Hello Dr Takeyuki, 

Thank you for your quick response. Unfortunately, your first suggestion did not work. 

At this point I need to understand if the sensor is broken or not. What would be the easiest way to check the UART output? Could it be an issue with my Raspbian OS?
